ThePhasesOfCraftship/1_code_first/assignment/begin/backend/src/index.ts

```diff
@@ -0,0 +1,186 @@
+    } catch (error) {
+        console.log('error handler')
+        next(error)
+    }
+});
+        
+        // Edit a user
+app.post('/users/edit/:userId', async (req: Request, res: Response, next: NextFunction) => {
+    // ...
+    const userId = parseInt(req.params.userId)
+    const { email, firstName, lastName, username } = req.body    
+    try {
+        const user = await updateUser(userId, { email, firstName, lastName, username })
+        if (!user) {
+            const userNotFoundResponse = responseObjectFactory(ErrorName.UserNotFound, undefined, false)
+            return res.status(404).json(userNotFoundResponse)
+        }
+        const response = responseObjectFactory(undefined, removePasswordForResponse(user), true)
+        res.status(200).send(response)
+    } catch (error) {
+        next(error)
+    }
+    
+    
+});
+
+// Get a user by email
+app.get('/users', async (req: Request, res: Response, next: NextFunction) => {
+    const email = req.query.email as string
+    try {
+        const user = await getUserByEmail(email)        
+        if (!user) {
+            const userNotFoundResponse = responseObjectFactory(ErrorName.UserNotFound, undefined, false)
+            return res.status(404).json(userNotFoundResponse)
+        }
+        const response = responseObjectFactory(undefined, removePasswordForResponse(user), true) 
+        res.status(200).send(response)
+    } catch (error) {
+        next(error)
+    }
+});
+
+app.use((error: Object, req: Request, res: Response, next: NextFunction) => {    
+    
+    if (error instanceof Prisma.PrismaClientValidationError) {
+        const validationErrorResponse = responseObjectFactory(ErrorName.ValidationError, undefined, false)
+        res.status(400).json(validationErrorResponse)
+    }
+    if (error instanceof Prisma.PrismaClientKnownRequestError) {            
+        if (error.code === 'P2002' && error.meta?.target && Array.isArray(error.meta?.target)) {
+            // DRY refactor error
+            if(error.meta.target.includes('email')){
+                const emailAlreadyInUseResponse = responseObjectFactory(ErrorName.EmailAlreadyInUse, undefined, false)
+                return res.status(409).json(emailAlreadyInUseResponse)
+            } 
+            if(error.meta.target.includes('username')) {
+                const usernameAlreadyTakenResponse = responseObjectFactory(ErrorName.UsernameAlreadyTaken, undefined, false)
+                return res.status(409).json(usernameAlreadyTakenResponse)
+            }
+            const clientErrorResponse = responseObjectFactory(ErrorName.ClientError, undefined, false) 
+            return res.status(400).json(clientErrorResponse);
+        }
+    }          
+    next(error) 
+})
+
+// Error handling middleware
+app.use((err, req: Request, res: Response) => {
+    const serverErrorResponse = responseObjectFactory(ErrorName.ServerError, undefined, false)
+    res.status(500).send(serverErrorResponse)
+})
```

Comment:
I'm a little hesitant with this approach. Please keep in mind that I'm a FE dev, with nearly no BE experience, so I'm just following my hunches :D 

What I see as a problem is that based on this code you rely on the DB to handle Application errors.
While I think aligning your DB schema with the business rules (unique email, username etc.) is correct, making the DB schema the "single source of truth" looks like a shortcut to me. Ofc this approach results in less code and sending less requests to the DB, but I think continuing down this path would result in scattering the business rules all over the place. Handling these errors in the application layer would be a cleaner approach IMO.
This approach also results in more coupling to the DB client, as you have to rely on Prisma's error types to construct your own errors.

Source: https://www.essentialist.dev/products/the-software-essentialist/categories/2154344001/posts/2168948146
<img width="653" alt="Screenshot 2024-08-01 at 16 12 09" src="https://github.com/user-attachments/assets/37d5fae8-7af3-43ff-84c6-aef7481fa19b">

Reply:
@enekesabel your comment is very good. Also, I think as a Code-First approach this seams. 
This middleware here functions as an adapter between the DB and the controller which is pretty good approach already at this stage (IMHO). This will be very easy to adapt when we evolve to richer architectural concepts down the road.